module04/03-Panda-Social-Network/solution.py

```diff
@@ -0,0 +1,151 @@
+import re
+
+
+class Panda:
+    """Panda which is user of :class:`PandaSocialNetwork` and has basic attributes as name,
+    email, gender and list of friends"""
+
+    def __init__(self, name, email, gender):
+        self.validate_input(name, email, gender)
+        self.panda_name = name
+        self.panda_email = email
+        self.panda_gender = gender
+        self.friend_list = []
+
+    def __str__(self):
+        return "%s panda named %s with email: %s" % (self.panda_gender.capitalize(), self.panda_name, self.panda_email)
+
+    def __eq__(self, other):
+        if isinstance(other, Panda):
+            return self.__key() == other.__key()
+        raise NotImplemented
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __key(self):
+        return self.panda_name, self.panda_email, self.panda_gender
+
+    def validate_input(self, name, email, gender):
+        if not name or not email or not gender:
+            raise ValueError("Input not provided")
+        if not self.is_valid_email(email):
+            raise ValueError("Not valid email provided")
+
+    @staticmethod
+    def is_valid_email(email):
+        email_regex = '^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$'
+        return re.search(email_regex, email)
+
+    def name(self):
+        return self.panda_name
+
+    def email(self):
+        return self.panda_email
+
+    def gender(self):
+        return self.panda_gender
+
+    def is_male(self):
+        return self.panda_gender == "male"
+
+    def is_female(self):
+        return self.panda_gender == "female"
+
+    def add_friend(self, panda):
+        self.friend_list.append(panda)
+
+    def friends(self):
+        return self.friend_list
+
+
+class PandaSocialNetwork:
+    """ Social network that can handle: adding of new :class:`Panda` users, making pandas friends,
+    checking connection level between pandas, checking how many pandas with specified gender
+    exist in pandas network"""
+
+    def __init__(self):
+        self.network = []
+
+    def add_panda(self, panda):
+        if self.has_panda(panda):
+            raise PandaAlreadyThere()
+        self.network.append(panda)
+
+    def has_panda(self, panda):
+        return panda in self.network
+
+    def make_friends(self, panda1, panda2):
+        if self.are_friends(panda1, panda2):
+            raise PandasAlreadyFriends
```

Comment:
you are missing `()` and therefore raising the class, not an instance of the exception

the tests are not failing, because we don't have a test for that - https://github.com/atodorov/qa-automation-python-selenium-101/issues/10